lib/src/test/java/io/ably/lib/test/realtime/EventEmitterTest.java

```diff
@@ -175,4 +176,30 @@ public void once_off_event_simple() {
 		assertNull(listener.counts.get(MyEvents.event_1));
 	}
 
+    @Test
```

Comment:
Did this test fail before you updated the code? Because if it adds a listener to the end of the collection, then I wouldn't expected it to have been called in the previous implementation.

In any case, please can there also be a test also that removes a listener, and verifies it is still called?

Reply:
The test never failed because there's no Listener added for MyEvents.event_1 in the test case. 
According to what I see, Listener is added for MyEvents.event_0, emitted for MyEvents.event_1, and then removed for MyEvents.event_0. 

So the test does pass before and after. 
I have added one test case `once_off_event` just to add listener, emit, remove listener and emit again.

Reply:
> The test never failed because there's no Listener added for MyEvents.event_1 in the test case.

OK, so how do you know that your change fixed anything, if you did not have a failing test? It's not always the case, but here I think, with knowledge of how things can fail with the old code, I think you could create a test that fails before the fix is made. Can you think of a way to do that?

Reply:
I tested this in code, but I might not have added a Junit test for this. I will add a test case as well and push commit. Thanks.

Reply:
OK, but here's a scenario that we know will fail with the old code, but will pass with the fixed code:

- add 2 listeners A, B, for an event;
- the listener for B removes A;
- call `emit()`;

With the old code B will be called twice, but with the fixed code A and B will each be called once.

Reply:
OK, well lets add a test for this removal case, which shows that you can remove a listener, from within the execution of another listener, without an exception. If we can we really have to have a test that shows the old behaviour is broken, and the new behaviour fixes it.